clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -71,10 +71,11 @@ void
 invariant_checker_t::report_if_false(per_shard_t *shard, bool condition,
                                      const std::string &invariant_name)
 {
+    uint64_t ref_count = shard->stream == nullptr ? shard->ref_count_
```

Comment:
This seems like it will mask errors: shard->ref_count_ is not guaranteed to be maintained (e.g., serial) and there would be no indication of a problem except a bogus count.  I think we should keep an invariant that the stream is present on entry here, and set it inside check_schedule_data which is already using a synthetic shard struct.

Reply:
I don't understand. This is not test code where we can create a test stream object. For a check done in `check_schedule_data` a bogus ref_count is fine since there is no ref count for a non-trace (cpu schedule) file.

Reply:
Is your concern about other cases where we do want a shard stream but don't have it. Shouldn't that be asserted elsewhere?

Reply:
Or do you mean change the report_if_false interface to accept a ref_count and tid.

Reply:
> Is your concern about other cases where we do want a shard stream but don't have it. Shouldn't that be asserted elsewhere?

Yes, that is a concern.  It does not seem good to allow stream to be empty but at the same time blindly assuming that ref_count_ is valid when we have cases where it's not (even if today we don't have cases where both are invalid...it just feels fragile).  I'm voting for:

- All `per_shard_t` types have a valid stream object.  If `check_schedule_data` wants to create a fake one, it has to fill in the stream, instead of forcing other functions that take `per_shard_t` to handle its incomplete shard.

Reply:
Just to be clear: your suggestion is to pass a test_memtrace_stream_t object from check_schedule_data (no overrides, just let the default get_record_ordinal return zero). This is doable but maybe we should rename test_memtrace_stream_t to default_memtrace_stream_t.

Reply:
Done.

Reply:
Looking at chedk_schedule_data it does update global.ref_count_: maybe that *should* that be part of the error report.

Reply:
Added.